subprojects/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java

```diff
@@ -46,16 +45,15 @@ private boolean isChangingValue(ProviderInternal<?> provider) {
 
     @Override
     protected Value<? extends R> calculateOwnValue(ValueConsumer consumer) {
-        Value<? extends A> lv = assertHasValue(left.calculateValue(consumer), left);
-        Value<? extends B> rv = assertHasValue(right.calculateValue(consumer), right);
-        return Value.of(combiner.apply(lv.get(), rv.get()));
-    }
-
-    private <T> Value<? extends T> assertHasValue(Value<? extends T> value, ProviderInternal<?> provider) {
-        if (value.isMissing()) {
-            throw new InvalidUserDataException("Provider has no value: " + provider);
+        Value<? extends A> lv = left.calculateValue(consumer);
+        Value<? extends B> rv = right.calculateValue(consumer);
```

Comment:
I think it would be best to avoid the `right.calculateValue` call when the left side is missing.

Reply:
Nice catch, thanks!